### PR TITLE
Look Up Table Approach for miniAOD packedCandidates

### PR DESCRIPTION
**Pull request by AdrianoDee**
##### Look Up Table Approach for _legacy_ miniAODs

This PR introduce the possibility to extract covariance elements from a Look Up Table for packedCandidates which do not have it stored. This is done through the `fallBackTrack()` method once the `setLut (int covV)` method is used on any of the packedCandidates in the collection.

The fallBackTrack() method works similarly to bestTrack() and returns a reco::Track pointer when available. When the PackedCandidate contains the track details (covariance matrix and number of hits) this method will be exacty the same as bestTrack(). When there are no track details inside the PackedCandidate (low pt candidates in legacy miniAODs for example), this method will use a look up table to estimate the covariance matrix elements based on the candidate pt, eta and number of total and pixel hits. Since the number of total and pixel hits are not stored for candidates with no track details, they are passed as arguments of fallBackTrack(). This method will return nullptr when no track was associated at production time (e.g. candidates with no track even in AOD).

Before using fallBackTrack for the first time the static method PackedCandidate::setLut(desired_version) should be called. This method will enable the use of the LUT and force the covariance packing method version for all candidates (default is to initialize it lazily at the first candidate). The fallBackTrack method will throw if this method is not called before.
It is important that the setLut method is called with the correct version for the miniAOD we are running (what is stored in PackedCandidates with track details). 

@elusian

This PR is completed by and depends on cms-data/DataFormats-PatCandidates#2 which adds a new schema, called 523, which always returns the plain LUT value.

#### PR Validation and Further references

BPH Jamboree December 2020 [here](https://indico.cern.ch/event/979184/contributions/4148171/attachments/2163278/3650677/MiniAOD_Run3_Adriano_JamboreeDic2020.pdf)

xPOG Meeting February 2021 [here](https://indico.cern.ch/event/994581/contributions/4181429/attachments/2196037/3712976/xPOG%20Run3%20MiniAOD%20BPH%20Feb2021.pdf)

Further reference [here](https://github.com/AdrianoDee/miniaodstudies/blob/main/README.md)

For a complete documentation, trk4bph meetings: [1](https://indico.cern.ch/event/983890/), [2](https://indico.cern.ch/event/975219/), [3](https://indico.cern.ch/event/968995/) , [4](https://indico.cern.ch/event/964780/), [5](https://indico.cern.ch/event/963424/), [6](https://indico.cern.ch/event/955820/)

### CODE REVIEW

**Comment by slava77** [comment]
assign xpog

File: DataFormats/PatCandidates/interface/PackedCandidate.h [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #include "DataFormats/VertexReco/interface/Vertex.h"
 #include "DataFormats/VertexReco/interface/VertexFwd.h"
 #include "FWCore/Utilities/interface/thread_safety_macros.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
```
Comment (slava77) [inline DataFormats/PatCandidates/interface/PackedCandidate.h:13]: this seems unnecessary
but if it is, follow #32859 to remove dependence on FWCore/MessageLogger in the header file here
```suggestion
```

**Comment by slava77** [comment]
> Code check has found code style and quality issues which could be resolved by applying following patch(s)

please update these so that further automated tests can run

**Comment by AdrianoDee** [comment]
please test

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
please abort

**Comment by mariadalfonso** [comment]
test parameters:

pull_request =  cms-data/DataFormats-PatCandidates#2

**Comment by mariadalfonso** [comment]
please test

**Comment by mbluj** [comment]
Hello,
I would add the `minPtForLowQualityTrackProperties` flag also to [PATLostTracks.cc](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATLostTracks.cc) as replacement of a hardcoded value, but probably w/o lowering the threshold below 0.5GeV (cost should be evaluated). It could be, however, a subject of a separate PR.

**Comment by slava77** [comment]
> This PR is completed by and depends on [cms-data/DataFormats-PatCandidates#2](https://github.com/cms-data/DataFormats-PatCandidates/pull/2) which adds a new schema, called 523, which always returns the plain LUT value.

@cms-sw/xpog-l2 @arizzi 
please remind me the numbering logic/convention for the covariance packing schemas.
(I couldn't find it in the earlier presentation https://indico.cern.ch/event/618375/contributions/2499908/attachments/1423225/2182008/miniaod-covariance-update.pdf )
Thank you.

**Comment by slava77** [comment]
@AdrianoDee 
is this feature intended to stay in 12X/master or is there a plan to have this in 106X UL bParking miniAOD?

**Comment by AdrianoDee** [comment]
The idea is to backport it to 106X for bParking yes. Shall I wait for this to be merged to backport?

**Comment by slava77** [comment]
> The idea is to backport it to 106X for bParking yes. Shall I wait for this to be merged to backport?

yes, let's converge here first

**Comment by slava77** [comment]
> DQMHistoTests: Total failures: 1429

I'd rather we integrate the fallback feature of PackedCandidate separately from the changes in the track pt cut in the PackedCandidateProducer.
This part is supposed to leave no differences.

May I suggest to update this PR with a value set to the old hardcoded default of minPtForLowQualityTrackProperties_.
Once we can establish with the tests that there are no differences we could get back to the non-zero value.

I find that the newly introduced parameters with default values in unpacking methods are confusing. 
- one, semantic argument, is that the arguments look generic, while they belong only to the fallback case -> the parameter names should be updated

Currently the dxyError and dzError methods are missing additional protections to check for use of LUT or hasTrackDetails to return consistent results (for a candidate without track details if `dxyError` is called before `fallBackTrack`, there will be a crash, while after it will return what's loaded by `fallBackTrack` call(ees)).

Do we have to have the option to vary these track-by-track? If `n = 8, int np = 3, int schema = 523` is valid for all tracks, then simply dealing with a fixed value in unpackCovarianceElement should be enough (a global choice for the fallback can be made when setLut is called). 
If `n` and `np` need to vary track by track by choice of an analyst, perhaps a better solution is to simply expose the track making methods and return a track to the analyst instead of loading the `track_` and `m_` into the data member of the collection. 

IIUC, the proposed solution with possibly variable `n, np, schema` is also not thread safe in some cases : concurrent calls to `fallBackTrack` with different values of `n, np, schema` will give non-reproducible and potentially inconsistent values of `track_` and `m_` data members.

A unique way of reading the data in a given process better be preserved.

**Comment by slava77** [comment]
What is the specific use case here and can the proposed solution be replaced with something more straightforward: 
1. cov/track recovery step: read packedCandidates and set a parameter(s) for cases that need to be recovered in a new updated collection
2. use the updated packed candidates directly

I would rather we have this solution instead of the global `setLut` magic.

**Comment by AdrianoDee** [comment]
> I'd rather we integrate the fallback feature of PackedCandidate separately from the changes in the track pt cut in the PackedCandidateProducer.
> This part is supposed to leave no differences.
> 
> May I suggest to update this PR with a value set to the old hardcoded default of minPtForLowQualityTrackProperties_.
> Once we can establish with the tests that there are no differences we could get back to the non-zero value.

Indeed this makes sense. I'll leave this PR for the LUT approach to keep track of this discussion and open a new one for the `minPtForLowQualityTrackProperties_ ` feature.

> I find that the newly introduced parameters with default values in unpacking methods are confusing.
> 
> * one, semantic argument, is that the arguments look generic, while they belong only to the fallback case -> the parameter names should be updated

You mean something like `numberOfHitsForLUT`  ?

> Currently the dxyError and dzError methods are missing additional protections to check for use of LUT or hasTrackDetails to return consistent results (for a candidate without track details if `dxyError` is called before `fallBackTrack`, there will be a crash, while after it will return what's loaded by `fallBackTrack` call(ees)).

Indeed, missed it, fixing those.

> Do we have to have the option to vary these track-by-track? If `n = 8, int np = 3, int schema = 523` is valid for all tracks, then simply dealing with a fixed value in unpackCovarianceElement should be enough (a global choice for the fallback can be made when setLut is called).

Two arguments for keeping the `n` and `np` configurable by the analyst: 

- For the present the analyst may be smart enough to have a better estimation of `n` and `np` given other factors (kinematics,geometry)
- For the future if, for storage reasons I suppose, we will need to store the candidate number of track hits but not the track covariance it could be useful to have it configurable. 

For the schema indeed could be defenitely fixed.

Just to understand the two solutions you propose here to avoid thread safety issues are:

>  If n and np need to vary track by track by choice of an analyst, perhaps a better solution is to simply expose the track making methods and return a track to the analyst instead of loading the track_ and m_ into the data member of the collection.

1. Exposing all the methods to return a the "fallBack" track instead of loading it internally. 

> What is the specific use case here and can the proposed solution be replaced with something more straightforward:
> 
> 1. cov/track recovery step: read packedCandidates and set a parameter(s) for cases that need to be recovered in a new updated collection
> 2. use the updated packed candidates directly
> 
> I would rather we have this solution instead of the global `setLut` magic.

2. Let the analyst create skimmed collection of candidates with the fallback tracked already loaded (that would bypass the possibility to set a different number of hits)

To leave the analyst the possibility to change the number of hits we would go for 1. What do you think?

**Comment by AdrianoDee** [comment]
Moved low pt threshold to #33777 (with the same for lostTracks @mbluj )

**Comment by slava77** [comment]
On 5/19/21 3:10 AM, AdrianoDee wrote:
>     I'd rather we integrate the fallback feature of PackedCandidate
>     separately from the changes in the track pt cut in the
>     PackedCandidateProducer.
>     This part is supposed to leave no differences.
> 
>     May I suggest to update this PR with a value set to the old
>     hardcoded default of minPtForLowQualityTrackProperties_.
>     Once we can establish with the tests that there are no differences
>     we could get back to the non-zero value.
> 
> Indeed this makes sense. I'll leave this PR for the LUT approach to keep
> track of this discussion and open a new one for the
> |minPtForLowQualityTrackProperties_ | feature.
> 

thanks


>     I find that the newly introduced parameters with default values in
>     unpacking methods are confusing.
> 
>       * one, semantic argument, is that the arguments look generic,
>         while they belong only to the fallback case -> the parameter
>         names should be updated
> 
> You mean something like |numberOfHitsForLUT| ?

yes, although considering it's 3 parameters, this makes the interface
pretty heavy

> 
>     Currently the dxyError and dzError methods are missing additional
>     protections to check for use of LUT or hasTrackDetails to return
>     consistent results (for a candidate without track details if
>     |dxyError| is called before |fallBackTrack|, there will be a crash,
>     while after it will return what's loaded by |fallBackTrack| call(ees)).
> 
> Indeed, missed it, fixing those.

OK

> 
>     Do we have to have the option to vary these track-by-track? If |n =
>     8, int np = 3, int schema = 523| is valid for all tracks, then
>     simply dealing with a fixed value in unpackCovarianceElement should
>     be enough (a global choice for the fallback can be made when setLut
>     is called).
> 
> Two arguments for keeping the |n| and |np| configurable by the analyst:
> 
>   * For the present the analyst may be smart enough to have a better
>     estimation of |n| and |np| given other factors (kinematics,geometry)
>   * For the future if, for storage reasons I suppose, we will need to
>     store the candidate number of track hits but not the track
>     covariance it could be useful to have it configurable.
> 
> For the schema indeed could be defenitely fixed.

uhm, having this need for variability gives more weight to avoid this
LUT solution and instead to try making an updated collection for analysis.

> 
> Just to understand the two solutions you propose here to avoid thread
> safety issues are:
> 
>     If n and np need to vary track by track by choice of an analyst,
>     perhaps a better solution is to simply expose the track making
>     methods and return a track to the analyst instead of loading the
>     track_ and m_ into the data member of the collection.

In the context of staying with just the PackedCandidate, yes.

However, as in the later comment, I gravitate to having a solution to
create new PackedCandidates.
[Some copy minimization could be to make a vector of pointers to
PackedCandidates with just the updated ones recreated while the
unchanged could stay the same.]


> 
>  1. Exposing all the methods to return a the "fallBack" track instead of
>     loading it internally.
> 
>     What is the specific use case here and can the proposed solution be
>     replaced with something more straightforward:
> 
>      1. cov/track recovery step: read packedCandidates and set a
>         parameter(s) for cases that need to be recovered in a new
>         updated collection
>      2. use the updated packed candidates directly
> 
>     I would rather we have this solution instead of the global |setLut|
>     magic.
> 
>  2. Let the analyst create skimmed collection of candidates with the
>     fallback tracked already loaded (that would bypass the possibility
>     to set a different number of hits)
> 
> To leave the analyst the possibility to change the number of hits we
> would go for 1. What do you think?


Wouldn't it be easier to use 2?
Just to clarify, I implied that in this case 2 there is no need for a
fallback track, the `bestTrack` interface will work out of the box.
Here I  think that having a single vector/collection with all
packedCandidates updated in a way that can be used uniformly would be
more straightforward to use rather than having a mix of "default" and
"fallback" collections.



> 
> —

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@AdrianoDee 
please clarify if you are going to make updates in this PR with the non-globla-LUT solution (solution1 and/or solution2 from the talk on Friday)?
Thank you.

**Comment by AdrianoDee** [comment]
@slava77 yes It's my plan. This is for the moment low-priority until the process of merging the PRs for the bParking remini is complete.

**Comment by slava77** [comment]
> @slava77 yes It's my plan. This is for the moment low-priority until the process of merging the PRs for the bParking remini is complete. Thanks!

please clarify on the current status.
It looks like the bParking remini PRs have settled

**Comment by slava77** [comment]
> > @slava77 yes It's my plan. This is for the moment low-priority until the process of merging the PRs for the bParking remini is complete. Thanks!
> 
> please clarify on the current status.
> It looks like the bParking remini PRs have settled

ping

**Comment by AdrianoDee** [comment]
Hi @slava77 sorry for the delay, I'll try to converge by the end of the week. Thanks

**Comment by slava77** [comment]
> Hi @slava77 sorry for the delay, I'll try to converge by next week. Thanks

@AdrianoDee 
please clarify on the status.
Thank you.

**Comment by AdrianoDee** [comment]
Hi @slava77, sorry for the delay. We could have find an alternative and much simpler solution: adding to `PackedCandidate.h` only the public method (basically a setter):

```
void setLUTParameters(int covSchema, int covarianceVersion, int nHits, int nPixelHits)
    {
      covarianceVersion_ = covarianceVersion;
      covarianceSchema_ = covSchema;
      packedHits_ = (nHits & trackPixelHitsMask) | (nPixelHits << trackStripHitsShift);
     
    }
```

to be used by the user on a candidate with no track details. Setting the number of hits this would allow the user to use `pseudoTrack` or `bestTrack`. We don’t see any drawback and have tested the method. 

Also, given this PR has diverged a bit and most of the commits would be useless I’d close it and open a new one referencing this for the discussion.

What do you think?

**Comment by AdrianoDee** [comment]
Worth to mention we would add a plugin to generate a new collection.

**Comment by slava77** [comment]
> adding to `PackedCandidate.h` only the public method (basically a setter):



> We could have find an alternative and much simpler solution: adding to `PackedCandidate.h` only the public method (basically a setter):
> 
> ```
> void setLUTParameters(int covSchema, int covarianceVersion, int nHits, int nPixelHits)
>     {
>       covarianceVersion_ = covarianceVersion;
> ```

I may still be missing something, but it seems like you are effectively interested in `void setTrackProperties(int nHits, int nPixelHits, int quality, int covarianceVersion)` (or `setTrackPropertiesLite`) which can be called without a full track

**Comment by AdrianoDee** [comment]
Yes, that would be a better naming probably.

**Comment by slava77** [comment]
@AdrianoDee 
please clarify on the status of this PR.
you mentioned earlier that this one will be closed.
Please close if that's the plan.
Thank you.

**Pull request closed without merging**